lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -0,0 +1,639 @@
+		$namespace_stack = array();
+		$context_stack   = array();
+		$unbalanced      = false;
+
+		$directive_processor_prefixes          = array_keys( self::$directive_processors );
+		$directive_processor_prefixes_reversed = array_reverse( $directive_processor_prefixes );
+
+		while ( $p->next_tag( array( 'tag_closers' => 'visit' ) ) && false === $unbalanced ) {
+			$tag_name = $p->get_tag();
+
+			if ( $p->is_tag_closer() ) {
+				list( $opening_tag_name, $directives_prefixes ) = end( $tag_stack );
+
+				if ( 0 === count( $tag_stack ) || $opening_tag_name !== $tag_name ) {
+					// If the tag stack is empty or the matching opening tag is not the
+					// same than the closing tag, it means the HTML is unbalanced and it
+					// stops processing it.
+					$unbalanced = true;
+					continue;
+				} else {
+					// It removes the last tag from the stack.
+					array_pop( $tag_stack );
+
+					// If the matching opening tag didn't have any directives, it can skip
+					// the processing.
+					if ( 0 === count( $directives_prefixes ) ) {
+						continue;
+					}
+				}
+			} else {
+				$directives_prefixes = array();
+
+				foreach ( $p->get_attribute_names_with_prefix( 'data-wp-' ) as $attribute_name ) {
+					// Extracts the directive prefix to see if there is a server directive
+					// processor registered for that directive.
+					list( $directive_prefix ) = $this->extract_prefix_and_suffix( $attribute_name );
+					if ( array_key_exists( $directive_prefix, self::$directive_processors ) ) {
+						$directives_prefixes[] = $directive_prefix;
+					}
+				}
+
+				// If this is not a void element, it adds it to the tag stack so it can
+				// process its closing tag and check for unbalanced tags.
+				if ( ! $p->is_void_element() ) {
+					$tag_stack[] = array( $tag_name, $directives_prefixes );
+				}
+			}
+
+			// Sorts the attributes by the order of the `directives_processor` array
+			// and checks what directives are present in this element. The processing
+			// order is reversed for tag closers.
+			$directives_prefixes = array_intersect(
+				$p->is_tag_closer()
+					? $directive_processor_prefixes_reversed
+					: $directive_processor_prefixes,
+				$directives_prefixes
+			);
+
+			// Executes the directive processors present in this element.
+			foreach ( $directives_prefixes as $directive_prefix ) {
+				$func = is_array( self::$directive_processors[ $directive_prefix ] )
+					? self::$directive_processors[ $directive_prefix ]
+					: array( $this, self::$directive_processors[ $directive_prefix ] );
+				call_user_func_array(
+					$func,
+					array( $p, &$context_stack, &$namespace_stack )
+				);
+			}
+		}
+
+		// It returns the original content if the HTML is unbalanced because
+		// unbalanced HTML is not safe to process. In that case, the Interactivity
+		// API runtime will update the HTML on the client side during the hydration.
+		return $unbalanced || 0 < count( $tag_stack ) ? $html : $p->get_updated_html();
```

Comment:
It's not worth changing this PR, but I think that this will currently fail (because of unbalanced tags) if processing SVG or MathML elements which are self-closing. We will be handling this in the HTML API soon (probably in 6.6) and so if it's the case we'll be able to lean on that functionality when it comes.

Reply:
I agree we can merge this and investigate that later so I've added a task to [the Tracking Issue](https://github.com/WordPress/gutenberg/issues/56803) to review what happens with SVG and MathML elements before WP 6.5.